css/css-view-transitions/paint-holding-in-iframe.html fails due to iframe not being composited

mattwoodrow opened the pull request:
#### 3ee3021ece1b24bad6071de534887d294b62aed9
<pre>
css/css-view-transitions/paint-holding-in-iframe.html fails due to iframe not being composited
<a href="https://bugs.webkit.org/show_bug.cgi?id=293468">https://bugs.webkit.org/show_bug.cgi?id=293468</a>

Reviewed by Simon Fraser.

Paint holding (suppression) requires a compositing layer for it to be applied,
and frames don&apos;t necessarily have one.

Add a condition to requiresCompositingForViewTransition so that iframe
RenderViews become composited if we&apos;re initializing a view-transition and will
try to suppress rendering. Add invalidation for this state to
Document::setActiveViewTransition.

Also fixes a race where rendering suppression is supposed to be applied at the
end of the rendering update, but instead happens after any async drawing is
completed (didCompleteRenderingFrame). This can mean the update callback runs
before rendering suppression starts.

Unify WebPage::didUpdateRendering/didPaintLayers so that we just call the one
function regardless of which DrawingArea impl is being used. Forward this
through to Page as the mark of the last main-thread piece of updateRendering,
and use that to enable rendering suppression.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setActiveViewTransition):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCompleteRenderingFrame):
(WebCore::Page::didUpdateRendering):
* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForViewTransition const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didUpdateRendering):
(WebKit::WebPage::didPaintLayers): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/295482@main">https://commits.webkit.org/295482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/061cf0d8ace76e4892b43774c0d7527a34630c63

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105220 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15357 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110434 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55882 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107261 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25360 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33476 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/79919 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108226 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19771 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94971 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60226 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13046 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55274 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/89230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13090 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113018 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32381 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88991 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32746 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91188 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88625 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22602 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33521 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11310 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27791 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32304 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/37718 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/32087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/35431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33652 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/19b9ac904fa3c9eecde08704860446678fbf568b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45811

nt1m commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2459,7 +2459,11 @@ void Page::didCompleteRenderingFrame()
 
+void Page::didUpdateRendering()
+{
+    LOG_WITH_STREAM(EventLoop, stream << "Page " << this << " didUpdateRendering");
```
```suggestion
    LOG_WITH_STREAM(EventLoop, stream << "Page " << this << " didUpdateRendering()");
```

smfr commented on Source/WebCore/page/Page.h:
```diff
@@ -816,11 +816,14 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     WEBCORE_EXPORT void didCompleteRenderingFrame();
+    // Called after the "display" steps of the rendering update, but before any async delays
+    // waiting for async painting.
+    WEBCORE_EXPORT void didUpdateRendering();
```
LegacyWebKit also needs to call this somewhere.

smfr reviewed (commented):
Typo in title.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ca97986077e959f475eb6584999267863b4be59e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45811

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/061cf0d8ace76e4892b43774c0d7527a34630c63. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45811

mattwoodrow committed:
css/css-view-transitions/paint-holding-in-iframe.html fails due to iframe not being composited
https://bugs.webkit.org/show_bug.cgi?id=293468

Reviewed by Simon Fraser.

Paint holding (suppression) requires a compositing layer for it to be applied,
and frames don't necessarily have one.

Add a condition to requiresCompositingForViewTransition so that iframe
RenderViews become composited if we're initializing a view-transition and will
try to suppress rendering. Add invalidation for this state to
Document::setActiveViewTransition.

Also fixes a race where rendering suppression is supposed to be applied at the
end of the rendering update, but instead happens after any async drawing is
completed (didCompleteRenderingFrame). This can mean the update callback runs
before rendering suppression starts.

Unify WebPage::didUpdateRendering/didPaintLayers so that we just call the one
function regardless of which DrawingArea impl is being used. Forward this
through to Page as the mark of the last main-thread piece of updateRendering,
and use that to enable rendering suppression.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setActiveViewTransition):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCompleteRenderingFrame):
(WebCore::Page::didUpdateRendering):
* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForViewTransition const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didUpdateRendering):
(WebKit::WebPage::didPaintLayers): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/295482@main
